internal/topology/variables/cluster_variable_validation.go

```diff
@@ -187,18 +224,61 @@ func validateUnknownFields(fldPath *field.Path, clusterVariable *clusterv1.Clust
 	return nil
 }
 
-func getClusterVariablesMap(clusterVariables []clusterv1.ClusterVariable) map[string]*clusterv1.ClusterVariable {
-	variablesMap := map[string]*clusterv1.ClusterVariable{}
-	for i := range clusterVariables {
-		variablesMap[clusterVariables[i].Name] = &clusterVariables[i]
+func getClusterVariablesMap(clusterVariables []clusterv1.ClusterVariable) (map[string]map[string]clusterv1.ClusterVariable, error) {
+	variablesMap := map[string]map[string]clusterv1.ClusterVariable{}
+	errs := []error{}
+	for _, clusterVariable := range clusterVariables {
+		c := clusterVariable
+		_, ok := variablesMap[c.Name]
+		if !ok {
+			variablesMap[c.Name] = map[string]clusterv1.ClusterVariable{}
+		}
+
+		// If the variable definitionFrom is empty, check that the variable has not been previously defined with a definitionFrom.
+		if ok && clusterVariable.DefinitionFrom == emptyVariableDefinitionFrom {
+			errs = append(errs, errors.Errorf("variable %q can not be defined with a mix of empty and non-empty values for definitionFrom", c.Name))
+			continue
+		}
+
+		// Check that the variable has not been defined more than once with the same definitionFrom.
+		if _, ok := variablesMap[c.Name][c.DefinitionFrom]; ok {
+			errs = append(errs, errors.Errorf("variable %q from %q is defined more than once", c.Name, c.DefinitionFrom))
+			continue
+		}
+
+		// Check that the variable has not also been defined without a definitionFrom
+		if _, ok := variablesMap[c.Name][emptyVariableDefinitionFrom]; ok {
+			errs = append(errs, errors.Errorf("variable %q can not be defined with a mix of empty and non-empty values for definitionFrom", c.Name))
+			continue
+		}
+
+		// Add the variable.
+		variablesMap[c.Name][c.DefinitionFrom] = c
 	}
-	return variablesMap
+	if len(errs) > 0 {
+		return nil, kerrors.NewAggregate(errs)
+	}
+	return variablesMap, nil
 }
 
-func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassVariable) map[string]*clusterv1.ClusterClassVariable {
-	variablesMap := map[string]*clusterv1.ClusterClassVariable{}
-	for i := range clusterClassVariables {
-		variablesMap[clusterClassVariables[i].Name] = &clusterClassVariables[i]
+func getClusterClassVariablesMap(clusterClassVariables []clusterv1.ClusterClassStatusVariable) map[ccVariableMapKey]statusVariableDefinition {
```

Comment:
Wondering if we should return the same data structure as above in getClusterVariablesMap => map[variable-name][definitionfrom]